### PR TITLE
[auth][bugfix] Import pvt keys with unknown file extension

### PR DESCRIPTION
**Pull request by elpaso**
This fixes an unreported bug that prevent imports of private keys with wrong/unknown extension.

The old logic relied on the file extension, that is  not only weak but also wrong because the same extension  can have different encodings.
    
The new implementation is 100% robust because completely  ignores the file extentions and try to load the key with  all supported encodings and algorithms before giving up.

With tests!

### CODE REVIEW

File: src/core/auth/qgsauthcertutils.cpp [hunk 1]
```diff
@@ -112,8 +112,6 @@ QByteArray QgsAuthCertUtils::fileData( const QString &path, bool astext )
   }
   // TODO: add checks for locked file, etc., to ensure it can be read
   QFile::OpenMode openflags( QIODevice::ReadOnly );
-  if ( astext )
-    openflags |= QIODevice::Text;
```
Comment (dakcarto) [inline src/core/auth/qgsauthcertutils.cpp]: Why is this removed? From `QIODevice::Text`:

"When reading, the end-of-line terminators are translated to '\n'. ..."

Don't we want that? Or, do you think the line endings will not matter in the `QByteArray`, if it is passed to the ctor of most `QSsl*` classes?

File: src/core/auth/qgsauthcertutils.cpp [hunk 2]
```diff
@@ -191,37 +189,55 @@ QSslKey QgsAuthCertUtils::keyFromFile( const QString &keypath,
...
+                       QSsl::PrivateKey,
+                       !keypass.isEmpty() ? keypass.toUtf8() : QByteArray() );
+  if ( ! clientkey.isNull() )
+  {
+    if ( algtype )
+      *algtype = QStringLiteral( "dsa" );
+    return clientkey;
+  }
+  return QSslKey();
 }
```
Comment (dakcarto) [inline src/core/auth/qgsauthcertutils.cpp]: We need to also test for `QSsl::Ec (The Elliptic Curve algorithm)`, though not sure how to test for `QSsl::Opaque` -- or not test for `QSsl::Opaque` and just return the empty QSslKey.

(see [QSsl::KeyAlgorithm](http://doc.qt.io/qt-5/qssl.html#KeyAlgorithm-enum))

File: src/core/auth/qgsauthcertutils.cpp (hunk 1 above)
Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: When reading as `QIODevice::Text` first, we can check for the following strings, which will let us know right away that the contents are PEM-base64 (from various output formats):

```
# PKCS#8 (algorithm is encoded in ASN.1 structure element)
-----BEGIN ENCRYPTED PRIVATE KEY
# Older OpenSSL formats, specific to algorithm used
-----BEGIN EC PRIVATE KEY
-----BEGIN RSA PRIVATE KEY
-----BEGIN DSA PRIVATE KEY
```

See final [PEM headers QSslKey expects](https://code.woboq.org/qt5/qtbase/src/network/ssl/qsslkey_p.cpp.html#_ZNK14QSslKeyPrivate9pemHeaderEv).

**Review by dakcarto — CHANGES_REQUESTED** [review]

File: src/core/auth/qgsauthcertutils.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: I don't think that reading the file in binary mode when it's ASCII gives a different byte array result than reading it in text mode. I made a quick test on a PEM key to verify that this is the case and all the pre-existing CI test cases are passing, that's why I removed that from the API.

Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: Do you think we should check if `-----BEGIN` is inside the payload and set the PEM type if true?
I don't see a problem with the current implementation (that is 100% correct) but adding the check would spare us some CPU cycles in most situations.

File: src/core/auth/qgsauthcertutils.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: Question is: if the type is Qssl::Opaque would it be treated differently than an empty/null key within QGIS?

File: src/core/auth/qgsauthcertutils.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: Done ^ 
The new implementation get the encoding from the file content and loop through all supported keys before giving up.

Added new test cases for EC and DSA keys, everything works as expected.

File: src/core/auth/qgsauthcertutils.cpp (hunk 2 above)
Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: I meant leave it as unsupported. However, best to leave how an opaque key is handled to Qt, since that is the point of that alg type, e.g. PKSC#11 smart card support, etc.

File: src/core/auth/qgsauthcertutils.cpp (hunk 1 above)
Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: New implementation looks good and clean.

My point about the `QIODevice::Text` reading is that it auto-converts line endings (otherwise, we might need to handle that). Since we have only test certs with Unix line endings, they should all pass. Looks like we need a PEM file with Win line endings to test. However, I'm not concerned with the read test or testing of the private key here, but elsewhere in the code where the read may _expect_ PEM text without odd line endings (I can't think of one right now - maybe the cert info dialog output, where the PEM is displayed).

Previously, we got the line ending conversion but only because we supposedly knew what way to open the file before calling the function. Maybe binary matching of "-----BEGIN " could prompt a re-read of the file as text. Not sure how important this is though.

Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: My concern here is that if an existing SSL public or private key file read from disk has Win line endings it will not cause issues during ctors of `QSslCertificate`, `QSslKey` or the similar counterparts in QCA. That might be a very hard issue to debug, i.e. the cert/key just appears to be broken, when in fact its just line endings causing problems. May not be an issue if all of those ctors manage different line endings internally, regardless of platform.

Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: I see, is there a way we can test this? But first: are you sure that a PEM cert encoded on windows will have cr+lf ? 
I hope that the internal implementation in QT is robust enough to parse the lines independently from the line ending. 
BTW, I can get the argument back.

Comment (elpaso) [reply src/core/auth/qgsauthcertutils.cpp]: @dakcarto I've added some more tests with CRLF and no line endings, all tests pass.

**Review by dakcarto — APPROVED** [review]
